pkg/storegateway/indexheader/reader_pool_test.go

```diff
@@ -28,54 +28,92 @@ import (
 
 func TestReaderPool_NewBinaryReader(t *testing.T) {
 	tests := map[string]struct {
-		lazyReaderEnabled     bool
-		lazyReaderIdleTimeout time.Duration
+		lazyReaderEnabled             bool
+		lazyReaderIdleTimeout         time.Duration
+		eagerLoadReaderEnabled        bool
+		persistLazyLoadedHeaderFn     func(blockId ulid.ULID) lazyLoadedHeadersSnapshot
+		expectedLoadCountMetricBefore int
+		expectedLoadCountMetricAfter  int
 	}{
 		"lazy reader is disabled": {
-			lazyReaderEnabled: false,
+			lazyReaderEnabled:            false,
+			expectedLoadCountMetricAfter: 0, // no lazy loading
 		},
 		"lazy reader is enabled but close on idle timeout is disabled": {
-			lazyReaderEnabled:     true,
-			lazyReaderIdleTimeout: 0,
+			lazyReaderEnabled:            true,
+			lazyReaderIdleTimeout:        0,
+			expectedLoadCountMetricAfter: 1,
 		},
 		"lazy reader and close on idle timeout are both enabled": {
-			lazyReaderEnabled:     true,
-			lazyReaderIdleTimeout: time.Minute,
+			lazyReaderEnabled:            true,
+			lazyReaderIdleTimeout:        time.Minute,
+			expectedLoadCountMetricAfter: 1,
+		},
+		"lazy reader preShutdownLoadedBlocks is present": {
+			lazyReaderEnabled:             true,
+			lazyReaderIdleTimeout:         time.Minute,
+			eagerLoadReaderEnabled:        true,
+			expectedLoadCountMetricBefore: 1, // the index header will be eagerly loaded before the operation
+			expectedLoadCountMetricAfter:  1,
+			persistLazyLoadedHeaderFn: func(blockId ulid.ULID) lazyLoadedHeadersSnapshot {
+				return lazyLoadedHeadersSnapshot{
+					IndexHeaderLastUsedTime: map[ulid.ULID]int64{blockId: time.Now().UnixMilli()},
+					UserID:                  "anonymous",
+				}
+			},
+		},
+		"no valid preShutdownLoadedBlocks is present": {
```

Comment:
Would something like "block is not present in pre-shutdown loaded blocks snapshot and eager-loading is enabled" be a clearer description of this test case?